server: don't drain when decommissioning

tbg opened the pull request:
Prior to this commit, a server which enters decommissioning would
automatically drain. This wasn't a great idea since it meant that
pointing the decommissioning command at a set of nodes required
for the cluster to work would brick that cluster, and would be
difficult to recover from (since the draining state is picked
up by the nodes when they start).

Instead, decouple draining from decommissioning. Decommissioning simply
tells the allocator to move data off that node; when the node gets shut
down cleanly (the operator's responsiblity), it will drain.

The resulting behavior when trying to decommission a too-large set of
nodes is now that the decommission command will simply not finish.
For example, starting a three node cluster and decommissioning all
nodes will simply hang (though the nodes will be marked as
decommissioning). Add three more nodes to the cluster and replicas
will move over to the newly added nodes, and the decommissioning
command will finish.

As a bonus, recommissioning a node now doesn't require the target
node to restart.

As a second bonus, the decommissioning acceptance test now takes
around 60% of the previous time (~45s down from 70+).

One remaining caveat is that users may forget that they attempted
to decommission nodes. We need to check that we prominently alert
in the UI when nodes are decommissioning. This isn't a new problem.

Fixes #27444.
Fixes #27025.

Release note (bug fix): decommissioning multiple nodes is now possible
without posing a risk to cluster health. Recommissioning a node does
no longer require a restart of the target nodes to take effect.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/28707)
<!-- Reviewable:end -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> but I think you mean "don't drain when decommissioning" in the commit/PR title.

> Reviewed 3 of 3 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28707#-:-LK2zszPADwbS836qemu:bjr7o91)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/acceptance/decommission_test.go, line 284 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/28707#-LK2zNrR80zEOboOTWqW:-LK2zNrR80zEOboOTWqX:buamo7) ([raw file](https://github.com/cockroachdb/cockroach/blob/89a285278e7451350404a83f1856ad4d139f5d47/pkg/acceptance/decommission_test.go#L284)):*
> ```Go
> 		exp := [][]string{
> 			decommissionHeader,
> 			// Expect the same usual, except this time the node should be draining
> ```

s/same usual/same as usual/


<!-- Sent from Reviewable.io -->

tbg committed:
server: don't drain when decommissioning

Prior to this commit, a server which enters decommissioning would
automatically drain. This wasn't a great idea since it meant that
pointing the decommissioning command at a set of nodes required
for the cluster to work would brick that cluster, and would be
difficult to recover from (since the draining state is picked
up by the nodes when they start).

Instead, decouple draining from decommissioning. Decommissioning simply
tells the allocator to move data off that node; when the node gets shut
down cleanly (the operator's responsiblity), it will drain.

The resulting behavior when trying to decommission a too-large set of
nodes is now that the decommission command will simply not finish.
For example, starting a three node cluster and decommissioning all
nodes will simply hang (though the nodes will be marked as
decommissioning). Add three more nodes to the cluster and replicas
will move over to the newly added nodes, and the decommissioning
command will finish.

As a bonus, recommissioning a node now doesn't require the target
node to restart.

As a second bonus, the decommissioning acceptance test now takes
around 60% of the previous time (~45s down from 70+).

One remaining caveat is that users may forget that they attempted
to decommission nodes. We need to check that we prominently alert
in the UI when nodes are decommissioning. This isn't a new problem.

Fixes #27444.
Fixes #27025.

Release note (bug fix): decommissioning multiple nodes is now possible
without posing a risk to cluster health. Recommissioning a node does
no longer require a restart of the target nodes to take effect.

tschottdorf commented:
Changed both, TFTR!

bors r=nvanbenschoten
